netcore/System.Private.CoreLib/src/System.Threading/Interlocked.cs

```diff
@@ -13,7 +13,26 @@ public static class Interlocked
 		public extern static int CompareExchange (ref int location1, int value, int comparand);
 
 		[MethodImplAttribute(MethodImplOptions.InternalCall)]
-		public extern static object CompareExchange (ref object location1, object value, object comparand);
+		extern static void CompareExchange_object (ref object location1, ref object value, ref object comparand, ref object result);
+
+		public static object CompareExchange (ref object location1, object value, object comparand)
+		{
+			// This avoids coop handles, esp. on the output which would be particularly inefficient.
+			// Passing everything by ref is equivalent to coop handles -- ref to locals at least.
+			//
+			// location1's treatment is unclear. But note that passing it by handle would be incorrect,
+			// as it would use a local alias, which the coop marshaling does, to avoid the unclarity here,
+			// that of a ref being to a managed frame vs. a native frame. Perhaps that could be revisited.
+			//
+			// So there a hole here, that of calling this function with location1 being in a native frame.
+			// Usually it will be to a field, static or not, and not even to managed stack.
+			//
+			// This is usually intrinsified. Ideally it is always intrinisified.
+			//
+			object result = null;
```

Comment:
Ideally the parameters and locals could be marked volatile. In that sense this is not equivalent to the icall wrapper. Maybe should fix that. And try “fixed” and address-of on the generic?